### PR TITLE
Finished assessment

### PR DESCRIPTION
**Pull request by ResetReesi**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [😎] JavaScript Coding Challenges
- [ 😎] Ruby Coding Challenges
- [ 😎] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:  Need to take my time and learn to refactor things. I realized that I have a problem with creating shorter code meaning putting everything into one function.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -11,40 +11,259 @@
...
 
-// b) Create the function that makes the test pass.
+// a) Create a test with an expect statement using the variable provided. 
+
+describe("colorsShuffle", () => {
+    it("removes the first item from the array and shuffles the remaining content", () => {
+        expect(colorsShuffle(colors1)).toEqual([ 'blue', 'green', 'pink', 'yellow' ])
+        expect(colorsShuffle2(colors2)).toEqual(['aquamarine', 'indigo', 'ochre', 'periwinkle', 'saffron' ])
+    })
+})
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:33]: Okay this is a good attempt and therefore acceptable. But your test already tells me we are on the wrong train tracks. The test should only be invoking a single function `colorsShuffle()` that takes in the parameters of the different colors each time it's invoked. 
So the goals are 
-> 1 function named colorsShuffle()
-> invoked 2 times
ex.
colorsShuffle(colors1)
colorsShuffle(colors1)
 ^ notice the invocation has the same name!

File: code-challenges.test.js [hunk 2]
```diff
@@ -11,40 +11,259 @@
...
-const votes2 = {upVotes: 2, downVotes: 33}
+// const votes2 = {upVotes: 2, downVotes: 33}
 // Expected output: -31
 
+// a) Create a test with expect statements for each of the variables provided.
+
+describe("mostPopularVotes", () => {
+    it("takes in an object that contains up votes and down votes and returns the net total of votes", () => {
+        expect(mostPopularVotes(votes1)).toEqual(11)
+        expect(mostPopularVotes2(votes2)).toEqual(-31)
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:109]: Alright we have the same problem going on here^

File: code-challenges.test.js [hunk 3]
```diff
@@ -11,40 +11,259 @@
...
+console.log(mostPopularVotes())
+
+
+const upVotes2 = 2
+const downVotes2= 33
+const votes2 = upVotes2 - downVotes2
+var mostPopularVotes2 = () => {
+    return votes2
+
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:168]: I like where your head is at but remeber these value are just test values. You are not going ot know what votes your users are giving you as a dev so you have to make code that works for all votes and considers the votes objects.

File: interview-questions.md [hunk 4]
```diff
@@ -10,40 +10,79 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 Your answer:
 
+Object-oriented programming is a a computer programming design that helps us developers organize our software designs based on computer data and it's objeacts, instead of functions and logic. The major differnce between object- oriented programming and functional programming is the way each program stores and manipulates its data.  
+
 Researched answer:
 
+Research taught me that you can manipulate data simply by storing it. For example in data forensics in order to maintain the data integity you have to leave the computer on and do not save any files because even truning off the computer can manipulate data. I have never delt with the FBI before, but I watch alot of FBI shows and the FBI never wants you to touch the computer. The FBI believes computer shouldnt be touched and left as is, because once its touched data has been manipulated. As we already know once we manipulate data it losses it's intergrity.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:17]: I love this answer -- however it doesn't feel related to OOP. I'd like to know what you learned about Object Oriented Programming in you research. OOP deals with encapsulation, inheritance, polymorphism, and abstraction.

 This is a great answer for data forensics.

File: interview-questions.md [hunk 5]
```diff
@@ -10,40 +10,79 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+
+
 Researched answer:
+ 
+ On my research journey I took a look into explicit returns and found out that explicit returns are the complete opposite of implicit returns. You will need a return in order for the tell a function ('Hey can you deliver this value") after you run the function. when coding its easy to recognized a explicit return, all you have to do is look for the word return after your function. A fun research fact I found out about implicit returns is that it can help you to write shorter code. You can tell the diffrence between inplict and explict with the => (arrow) synatax. An implict return is noticed by the => syntax. Implict function use the arrow syntax simply to replace the traditional function. I have just one more fun fact the about implicts returns. The => (arrow) syntax has many built in meanings, while doing my research I found that pretty cool.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:40]: Excellent answer!

File: interview-questions.md [hunk 6]
```diff
@@ -10,40 +10,79 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+
+
 
 4. What is a block in Ruby?
 
 Your answer:
 
+A ruby block is no different from any other code block, it's pretty much a group of code. However some of the techniques may differ. for example A ruby block usually starts with the do and end with the word end. Do is were the code starts and end is where the code ends. It also can keep code enclosed the code with {} curly braces. Do and End is usually used for multiple lines of code whereas the {} curly braces is used if your workin with one line of code. All blocks have arguments but the ruby block takes in an argument and that aregument should be defined in between the pipes ||. Ruby blocks do not have to be identified by any name, but you must always pass a method call.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:49]: Nice!

File: interview-questions.md [hunk 7]
```diff
@@ -10,40 +10,79 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 Researched answer:
 
+Research shows that there is really no diffrence between a ruby block and a method however ruby blocks do not belong to a object. There are only 2 ways to declare a rubly block. Using the DO and End keywords is how ruby blocks are determined. Do is usually placed at the begginig of your block to be repeatedly executes. End it just where you would like your block of code to end. 
+
+
 5. What is an instance variable in Ruby?
 
 Your answer:
 
+If you ever recognized a variable that start with an @symbol we call this an instance variable in Ruby. Most instance variables are used to give objects there own land to store its data. Ruby instance variable are typically associates with objects and everything in ruby is an object.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:61]: love it!

File: ruby-challenges.rb [hunk 8]
```diff
@@ -23,12 +38,36 @@
...
+  if string.downcase == string.reverse.downcase
+    p 'is a palindrome'
+  else
+    p 'is not a palindrome'
+  end
+end
+
+palindromeStrings palindrome_tester1
+palindromeStrings palindrome_tester2
+palindromeStrings palindrome_tester3
```
Comment (WalkerTexas-coder) [inline ruby-challenges.rb:73]: These are great examples of writing out functions once and invoking it multiple times!

**Review by WalkerTexas-coder — APPROVED** [review]
Okay Brook, your ruby code and your interview questions area top notch stuff! I know you struggled with the stress of this last week and I appreciate the effort you put into this. There is one concept I want you to get down moving forward and that's how to write and code out javascript functions without reworking the input data. Each problem should have a single function as it's answer and a good function should handle most input situations - not a single input situations. That said, this is approved! 

Checkout the syllabus page on function is unit 1: javascript for help! 
https://github.com/learn-academy-2022-echo/Syllabus/blob/main/javascript/functions.md#pseudocode

**Comment by ResetReesi** [comment]
I agree Austin, I definitely need to study some JavaScript functions.
Totally agree 💯 with you on that one. I am working on it . Thanks I will
be going over the syllabus for help !!

**Pull request merged**